### PR TITLE
Fixes #20957 - Replace alias_method_chain with Module prepend

### PR DESCRIPTION
**Pull request by mmoll**
Deprecated in Rails 5.0 and will be removed in 5.1. Some instances of
classes overwriting existing methods can be handled with `super`, other
concerns or modules are changed to use prepend instead of include.

Note: no ActiveSupport::Concern support for prepends, so load class
methods the pure Ruby way.

### CODE REVIEW

**Comment by mmoll** [comment]
taken from Dominic's rails51 branch

File: test/controllers/hosts_controller_test.rb
Comment (mmoll) [inline test/controllers/hosts_controller_test.rb]: bit unsure how to handle this.

File: app/models/concerns/fog_extensions/openstack/server.rb
Comment (timogoebel) [inline app/models/concerns/fog_extensions/openstack/server.rb:4]: How about creating a custom `ActiveSupport::Concern` implementation so you can do

```ruby
prepended do

end
```

?

Comment (mmoll) [reply app/models/concerns/fog_extensions/openstack/server.rb:4]: you mean something similar to https://gitlab.com/gitlab-org/gitlab-ee/blob/master/config/initializers/0_as_concern.rb ?

Comment (timogoebel) [reply app/models/concerns/fog_extensions/openstack/server.rb:4]: Yeah. Exactly. But should be a separate PR I think.

Comment (mmoll) [reply app/models/concerns/fog_extensions/openstack/server.rb:4]: What's the opinion of other core devs here? Personally I'd rather like to see this PR get in now to speed things up, but no hard feelings either way.

Comment (timogoebel) [reply app/models/concerns/fog_extensions/openstack/server.rb:4]: I agree. My suggestion should not block this PR at all.

File: app/controllers/concerns/foreman/controller/smart_proxy_auth.rb
Comment (iNecas) [inline app/controllers/concerns/foreman/controller/smart_proxy_auth.rb]: It seems we don't need a dynamic anonymous module now, as there is nothing in the module block that would require a closure. I would prefer having standard module (something like `SmartProxyRequireSsl`) that would define this method and we would use that prepending. If nothing else, it's easier to see what modules are extending the classes (compating to anonymous modules)

File: app/controllers/unattended_controller.rb
Comment (iNecas) [inline app/controllers/unattended_controller.rb:10]: Couldn't we use:

```ruby
skip_before_action :require_login, :session_expiry, :update_activity_time, :set_taxonomy, :authorize, :unless => Proc.new { preview? }
```

File: app/models/subnet.rb
Comment (iNecas) [inline app/models/subnet.rb:284]: Wasn't there the `new_without_cast` for a reason there? Why wasn't there just `Subnet::Ipv4.new(*attributes, &block)` before?

File: test/controllers/hosts_controller_test.rb
Comment (iNecas) [reply test/controllers/hosts_controller_test.rb]: Where does the `save` call comes from?

**Review by iNecas — CHANGES_REQUESTED** [review]

File: app/models/subnet.rb
Comment (mmoll) [reply app/models/subnet.rb:284]: @timogoebel do you recall your original reason? :)

File: test/controllers/hosts_controller_test.rb
Comment (mmoll) [reply test/controllers/hosts_controller_test.rb]: app/models/concerns/foreman/sti.rb:26

File: app/controllers/unattended_controller.rb
Comment (mmoll) [reply app/controllers/unattended_controller.rb:10]: seems to work 👍

File: app/models/subnet.rb
Comment (mmoll) [reply app/models/subnet.rb:284]: I think that's also because of the change in app/models/concerns/foreman/sti.rb

File: app/controllers/concerns/foreman/controller/smart_proxy_auth.rb
Comment (mmoll) [reply app/controllers/concerns/foreman/controller/smart_proxy_auth.rb]: I'm not fully getting it, could you lend me a hand there? ✋

File: app/models/subnet.rb
Comment (iNecas) [reply app/models/subnet.rb:284]: As long as the tests are green, I'm ok with that, just wondering if there isn't something hidden around this.

File: app/controllers/concerns/foreman/controller/smart_proxy_auth.rb
Comment (iNecas) [reply app/controllers/concerns/foreman/controller/smart_proxy_auth.rb]: You would define separate module next to the `ClassMethods` like this;

```ruby
     module SmartProxyRequireSsl
        def require_ssl?
          if [self.smart_proxy_filter_actions].flatten.map(&:to_s).include?(self.action_name)
            false
          else
            super
          end
        end
      end
```
And then, here you would just do `prepend SmartProxyRequireSsl`

File: test/controllers/hosts_controller_test.rb
Comment (iNecas) [reply test/controllers/hosts_controller_test.rb]: So the two tests that have this issue are https://github.com/theforeman/foreman/blob/develop/test/controllers/hosts_controller_test.rb?utf8=%E2%9C%93#L1324 and https://github.com/theforeman/foreman/blob/develop/test/controllers/hosts_controller_test.rb?utf8=%E2%9C%93#L1336 and it's just because before this change, the `save` call would not be detected, because the actual method was `save_with_type` :D So we were actually calling the save there, and this expectation was not just catching it.

I think the issue was introduced by refactoring in https://github.com/theforeman/foreman/commit/b3fb472c60eeeb085859de58d771a5fec7de7ba3, where the expects were extracted to before filter without realizing there are some tests that actually call save (and we were not hitting issues just because of the aliasing).

We should move the expects just to the tests that really expect nothing to be saved: the explicitness would also help with readability of the expacations IMO

Comment (mmoll) [reply test/controllers/hosts_controller_test.rb]: Shouldn't the stuff from `describe '#template_used'` be isolated from the `test_submit_multiple_rebuild_config_*` tests?

Comment (iNecas) [reply test/controllers/hosts_controller_test.rb]: Yes, it seems those two are actually not related to the rest of the context. +1 for getting them out of `template_used`

Comment (mmoll) [reply test/controllers/hosts_controller_test.rb]: These two are not inside that `describe` block, so I have no idea why the `expects(:save).never` is triggered there

Comment (mmoll) [reply test/controllers/hosts_controller_test.rb]: ok, I think I got this...

**Comment by mmoll** [comment]
:green_heart:

**Review by iNecas — APPROVED** [review]

**Comment by iNecas** [comment]
Thanks @mmoll

**Pull request merged**
